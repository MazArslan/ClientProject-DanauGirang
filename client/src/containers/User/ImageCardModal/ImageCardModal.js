import React, { useState, useEffect, useCallback } from "react";
import { useHistory } from "react-router-dom";
import {
  Container,
  Row,
  Col,
  Modal,
  ModalHeader,
  ModalBody,
  Button
} from "reactstrap";
import propTypes from "prop-types";
import { faLink } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import CaptureEventForm from "../CaptureEventForm/CaptureEventForm";
import PaginationButton from "../../../components/UI/PaginationButton/PaginationButton";
import "./ImageCard.css";

/**
 * ImageCard component
 * Handles the capture event form modal functionality
 * @param { 
    cardId,
    projectId,
    tags,
    totalCaptureEvents,
    pageNumber,
    pageSize,
    captureEventsId,
    isOpen,
    toggle } props
 */
const ImageCardModal = props => {
  const {
    cardId,
    projectId,
    tags,
    totalCaptureEvents,
    pageNumber,
    pageSize,
    captureEventsId,
    isOpen,
    toggle
  } = props;
  const [currentCardId, setCurrentCardId] = useState();
  const [captureEventId, setCaptureEventId] = useState();
  const [captureEventIndex, setCaptureEventIndex] = useState();
  const [forwardIsDisabled, setForwardIsDisabled] = useState();
  const [backIsDisabled, setBackIsDisabled] = useState();
  const history = useHistory();

  const retrieveCaptureEventId = (cardNo, pageNo, size, captureEvents) => {
    // Calculates the index of the selected card in relation to it's index within the current page
    const pageIndex = pageNo - 1;
    const numPriorCapEvents = pageIndex * size;
    const index = cardNo - numPriorCapEvents - 1;
    setCaptureEventIndex(index);
    const id = captureEvents[index];
    return id;
  };

  // Sets the current card as the selected image grid thumbnail
  useEffect(() => {
    setCurrentCardId(cardId);
  }, [cardId]);

  useEffect(() => {
    const retrievedCaptureEvent = retrieveCaptureEventId(
      currentCardId,
      pageNumber,
      pageSize,
      captureEventsId
    );
    setCaptureEventId(retrievedCaptureEvent);
  }, [currentCardId, pageNumber, pageSize, captureEventsId]);

  // Handles enabling and disabling the buttons used to navigate the image cards
  useEffect(() => {
    const cardNumber = captureEventIndex + 1;
    const value = totalCaptureEvents < pageSize ? totalCaptureEvents : pageSize;
    if (value === 1) {
      setForwardIsDisabled(true);
      setBackIsDisabled(true);
    } else if (cardNumber >= value) {
      setForwardIsDisabled(true);
      setBackIsDisabled(false);
    } else if (cardNumber <= 1) {
      setForwardIsDisabled(false);
      setBackIsDisabled(true);
    } else if (cardNumber <= value && cardNumber >= 1) {
      setForwardIsDisabled(false);
      setBackIsDisabled(false);
    }
  }, [captureEventIndex, pageSize, totalCaptureEvents]);

  // Triggered when confirmation modal submit button clicked
  const modalSubmissionHandler = useCallback(
    e => {
      e.preventDefault();
      const nextCard = currentCardId + 1;
      if (nextCard <= totalCaptureEvents) {
        setCurrentCardId(nextCard);
      }
    },
    [currentCardId, totalCaptureEvents]
  );

  // Pagination bar arrow click handler
  const handleArrowClick = useCallback((e, index) => {
    e.preventDefault();
    setCurrentCardId(index);
  }, []);

  let cardTitle = null;
  if (backIsDisabled !== undefined && forwardIsDisabled !== undefined) {
    cardTitle = (
      <Row>
        <Col className="text-left">
          <PaginationButton
            isDisabled={backIsDisabled}
            clicked={event => handleArrowClick(event, currentCardId - 1)}
            text={"<"}
          />
        </Col>
        <Col className="text-center">
          <h2>{`Image ${currentCardId}`}</h2>
        </Col>
        <Col className="text-right">
          <PaginationButton
            isDisabled={forwardIsDisabled}
            clicked={event => handleArrowClick(event, currentCardId + 1)}
            text={">"}
          />
        </Col>
      </Row>
    );
  }

  let form = null;
  if (captureEventId && currentCardId) {
    form = (
      <CaptureEventForm
        tags={tags}
        captureEventId={captureEventId}
        submitted={e => modalSubmissionHandler(e)}
      />
    );
  }

  const permalinkHandler = () => {
    history.push(`/project/${projectId}/capture-event/${captureEventId}/form`);
  };

  const linkIcon = (
    <Row>
      <Col className="text-centre" xs="auto" style={{ paddingRight: "8px" }}>
        <FontAwesomeIcon icon={faLink} className="link-icon" />
      </Col>
      <Col style={{ paddingLeft: "0px" }}>
        <p>Permalink</p>
      </Col>
    </Row>
  );

  return (
    <Modal style={{ maxWidth: "900px" }} isOpen={isOpen} toggle={toggle}>
      <ModalHeader toggle={toggle} />
      <ModalBody>
        <React.Fragment>
          <div className="image-card">
            <Container>{cardTitle}</Container>
            <Container>
              <Row>
                <Col>
                  <Button
                    color="link"
                    onClick={() => permalinkHandler()}
                    className="permalink-button"
                  >
                    {linkIcon}
                  </Button>
                </Col>
              </Row>
            </Container>
            {form}
          </div>
        </React.Fragment>
      </ModalBody>
    </Modal>
  );
};

ImageCardModal.propTypes = {
  cardId: propTypes.number.isRequired,
  totalCaptureEvents: propTypes.number.isRequired,
  projectId: propTypes.string.isRequired,
  tags: propTypes.array.isRequired,
  captureEventsId: propTypes.array.isRequired,
  pageNumber: propTypes.number.isRequired,
  pageSize: propTypes.number.isRequired,
  isOpen: propTypes.bool.isRequired,
  toggle: propTypes.func.isRequired
};

export default ImageCardModal;
