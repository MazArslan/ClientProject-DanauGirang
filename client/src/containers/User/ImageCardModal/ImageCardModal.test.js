import React from "react";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";
import ImageCardModal from "./ImageCardModal";

Enzyme.configure({ adapter: new Adapter() });

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({})
}));

describe("<ImageCardModal/>", () => {
  const onActionMock = jest.fn();
  const props = {
    captureEventsId: ["CaptureEventId1", "CaptureEventId2", "CaptureEventId3"],
    projectId: "ProjectId1",
    cardId: 2,
    tags: ["Tiger", "Elephant", "Tapir"],
    totalCaptureEvents: 10,
    pageNumber: 2,
    pageSize: 6,
    isOpen: false,
    toggle: onActionMock
  };

  describe("component rendering", () => {
    const wrapper = shallow(<ImageCardModal {...props} />);

    it("should render shallow correctly with props", () => {
      expect(wrapper.exists()).toBe(true);
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("should render permalink button", () => {
      const permalinkButton = wrapper.find(
        'Button[className="permalink-button"]'
      );
      expect(permalinkButton.props().className).toEqual("permalink-button");
      expect(permalinkButton.props().tag).toEqual("button");
    });

    it("should render permalink button contents", () => {
      const buttonText = wrapper.find("p");
      expect(buttonText.props().children).toEqual("Permalink");
    });
  });
});
