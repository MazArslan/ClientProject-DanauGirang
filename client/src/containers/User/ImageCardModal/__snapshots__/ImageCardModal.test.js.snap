// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageCardModal/> component rendering should render shallow correctly with props 1`] = `
<Modal
  autoFocus={true}
  backdrop={true}
  backdropTransition={
    Object {
      "mountOnEnter": true,
      "timeout": 150,
    }
  }
  centered={false}
  fade={true}
  isOpen={false}
  keyboard={true}
  modalTransition={
    Object {
      "timeout": 300,
    }
  }
  onClosed={[Function]}
  onOpened={[Function]}
  returnFocusAfterClose={true}
  role="dialog"
  scrollable={false}
  style={
    Object {
      "maxWidth": "900px",
    }
  }
  toggle={[MockFunction]}
  unmountOnClose={true}
  zIndex={1050}
>
  <ModalHeader
    charCode={215}
    closeAriaLabel="Close"
    tag="h5"
    toggle={[MockFunction]}
    wrapTag="div"
  />
  <ModalBody
    tag="div"
  >
    <div
      className="image-card"
    >
      <Container
        tag="div"
      />
      <Container
        tag="div"
      >
        <Row
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
        >
          <Col
            tag="div"
            widths={
              Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ]
            }
          >
            <Button
              className="permalink-button"
              color="link"
              onClick={[Function]}
              tag="button"
            >
              <Row
                tag="div"
                widths={
                  Array [
                    "xs",
                    "sm",
                    "md",
                    "lg",
                    "xl",
                  ]
                }
              >
                <Col
                  className="text-centre"
                  style={
                    Object {
                      "paddingRight": "8px",
                    }
                  }
                  tag="div"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                  xs="auto"
                >
                  <FontAwesomeIcon
                    border={false}
                    className="link-icon"
                    fixedWidth={false}
                    flip={null}
                    icon={
                      Object {
                        "icon": Array [
                          512,
                          512,
                          Array [],
                          "f0c1",
                          "M326.612 185.391c59.747 59.809 58.927 155.698.36 214.59-.11.12-.24.25-.36.37l-67.2 67.2c-59.27 59.27-155.699 59.262-214.96 0-59.27-59.26-59.27-155.7 0-214.96l37.106-37.106c9.84-9.84 26.786-3.3 27.294 10.606.648 17.722 3.826 35.527 9.69 52.721 1.986 5.822.567 12.262-3.783 16.612l-13.087 13.087c-28.026 28.026-28.905 73.66-1.155 101.96 28.024 28.579 74.086 28.749 102.325.51l67.2-67.19c28.191-28.191 28.073-73.757 0-101.83-3.701-3.694-7.429-6.564-10.341-8.569a16.037 16.037 0 0 1-6.947-12.606c-.396-10.567 3.348-21.456 11.698-29.806l21.054-21.055c5.521-5.521 14.182-6.199 20.584-1.731a152.482 152.482 0 0 1 20.522 17.197zM467.547 44.449c-59.261-59.262-155.69-59.27-214.96 0l-67.2 67.2c-.12.12-.25.25-.36.37-58.566 58.892-59.387 154.781.36 214.59a152.454 152.454 0 0 0 20.521 17.196c6.402 4.468 15.064 3.789 20.584-1.731l21.054-21.055c8.35-8.35 12.094-19.239 11.698-29.806a16.037 16.037 0 0 0-6.947-12.606c-2.912-2.005-6.64-4.875-10.341-8.569-28.073-28.073-28.191-73.639 0-101.83l67.2-67.19c28.239-28.239 74.3-28.069 102.325.51 27.75 28.3 26.872 73.934-1.155 101.96l-13.087 13.087c-4.35 4.35-5.769 10.79-3.783 16.612 5.864 17.194 9.042 34.999 9.69 52.721.509 13.906 17.454 20.446 27.294 10.606l37.106-37.106c59.271-59.259 59.271-155.699.001-214.959z",
                        ],
                        "iconName": "link",
                        "prefix": "fas",
                      }
                    }
                    inverse={false}
                    listItem={false}
                    mask={null}
                    pull={null}
                    pulse={false}
                    rotation={null}
                    size={null}
                    spin={false}
                    swapOpacity={false}
                    symbol={false}
                    title=""
                    transform={null}
                  />
                </Col>
                <Col
                  style={
                    Object {
                      "paddingLeft": "0px",
                    }
                  }
                  tag="div"
                  widths={
                    Array [
                      "xs",
                      "sm",
                      "md",
                      "lg",
                      "xl",
                    ]
                  }
                >
                  <p>
                    Permalink
                  </p>
                </Col>
              </Row>
            </Button>
          </Col>
        </Row>
      </Container>
    </div>
  </ModalBody>
</Modal>
`;
